internal/handler/handler.go

```diff
@@ -246,69 +246,61 @@ func (h *Handler) cleanupAllCappedCollections(ctx context.Context) error {
 func (h *Handler) cleanupCappedCollection(ctx context.Context, db backends.Database, cInfo *backends.CollectionInfo, force bool) (int32, int64, error) { //nolint:lll // for readability
 	must.BeTrue(cInfo.Capped())
 
+	var docsDeleted int32
+	var bytesFreed int64
+	var statsBefore, statsAfter *backends.CollectionStatsResult
+
 	coll, err := db.Collection(cInfo.Name)
 	if err != nil {
 		return 0, 0, lazyerrors.Error(err)
 	}
 
-	statsBefore, err := coll.Stats(ctx, &backends.CollectionStatsParams{Refresh: true})
+	statsBefore, err = coll.Stats(ctx, &backends.CollectionStatsParams{Refresh: true})
 	if err != nil {
 		return 0, 0, lazyerrors.Error(err)
 	}
-
 	h.L.Debug("cleanupCappedCollection: stats before", zap.Any("stats", statsBefore))
 
-	if statsBefore.SizeCollection < cInfo.CappedSize && statsBefore.CountDocuments < cInfo.CappedDocuments {
-		return 0, 0, nil
-	}
-
-	res, err := coll.Query(ctx, &backends.QueryParams{
-		Sort:          must.NotFail(types.NewDocument("$natural", int64(1))),
-		Limit:         int64(float64(statsBefore.CountDocuments) * float64(h.CappedCleanupPercentage) / 100),
-		OnlyRecordIDs: true,
-	})
-	if err != nil {
-		return 0, 0, lazyerrors.Error(err)
-	}
-
-	defer res.Iter.Close()
-
-	var recordIDs []int64
-	for {
-		var doc *types.Document
-		if _, doc, err = res.Iter.Next(); err != nil {
-			if errors.Is(err, iterator.ErrIteratorDone) {
-				break
-			}
+	if count := getDocCleanupCount(cInfo, statsBefore); count > 0 {
+		err = deleteFirstNDocuments(ctx, coll, count)
+		if err != nil {
+			return 0, 0, lazyerrors.Error(err)
+		}
 
+		statsAfter, err = coll.Stats(ctx, &backends.CollectionStatsParams{Refresh: true})
+		if err != nil {
 			return 0, 0, lazyerrors.Error(err)
 		}
 
-		recordIDs = append(recordIDs, doc.RecordID())
-	}
+		h.L.Debug("cleanupCappedCollection: stats after document count reduction", zap.Any("stats", statsAfter))
+
+		docsDeleted += int32(count)
+		bytesFreed += (statsBefore.SizeTotal - statsAfter.SizeTotal)
 
-	if len(recordIDs) == 0 {
-		h.L.Debug("cleanupCappedCollection: no documents to delete")
-		return 0, 0, nil
+		statsBefore = statsAfter
 	}
 
-	deleteRes, err := coll.DeleteAll(ctx, &backends.DeleteAllParams{RecordIDs: recordIDs})
-	if err != nil {
-		return 0, 0, lazyerrors.Error(err)
+	if count := getSizeCleanupCount(cInfo, statsBefore, h.CappedCleanupPercentage); count > 0 {
+		err = deleteFirstNDocuments(ctx, coll, count)
```

Comment:
At the moment, `DeleteAll` document could be called twice, if there are too many document which deletes some, then if the collection sizes are too big it calls delete again. Calling delete requires postgresql delete operation so it would be better if we only call it once to minimize overhead :hugs:

I think it should be simple to do, call `getDocCleanupCount()` and `getSizeCleanupCount()`, then use which ever is bigger count.

Reply:
This is done so that we will be more precise w.r.t how many documents we are dropping. If both doc count and collection size exceeds the threshold, first we delete docs based on count, check the stats again and if collection size is within the limit, then we don't take further action.

> call getDocCleanupCount() and getSizeCleanupCount(), then use which ever is bigger count.

I think this will lead to the old behaviour, where there are chances of deleting docs unnecessarily (as pointed out in my `1M` example above).

Reply:
Ah I see your point. So your point is it has to delete by document count to then get the statistics to avoid deleting too many. Could you document that in the code please :pray: 

FYI I think we need to update `bytesFreed` after your second delete.

Reply:
1. Documented the code.
2. We don't collect the stats immediately after second deletion. `bytesFreed` is updated after compact and after collecting the stats.